File: web/src/types/cotizacion.ts
export interface MaterialItem {
  id?: string
  cat: string
  ref: string
  precio_m2: number
  area_placa: number
  largo?: number
  ancho?: number
  cantLaminas?: number
}

export interface PiezaItem {
  nombre: string
  ml: number
  ancho_custom: number
  cantidad: number
  categoria: string
  unidad_venta: string
  placa_idx?: number
}

export interface CotizacionDirectaIn {
  // Material
  categoria: string
  referencia: string
  precio_m2: number
  area_placa_comprada: number
  materiales_lista: MaterialItem[]
  piezas: PiezaItem[]
  // Proyecto
  tipo_proyecto: string
  etapa_label: string
  nombre_cliente: string
  margen_pct: number
  dias: number
  personas: number
  zocalo_activo: boolean
  zocalo_ml: number

  incluir_iva: boolean
}

export interface CotizacionResult {
  precio_sugerido: number
  costo_total: number
  utilidad: number
  margen_pct: number
  aprovechamiento: number
  c1_material: number
  c2_mano_obra: number
  c3_zocalos: number
  c4_insumos: number
  c5_logistica: number
  c6_viaticos: number
  c7_adicionales: number
  retal: number
  m2_real: number
  categoria: string
}

// ── AIU ──────────────────────────────────────────────────────────────────────

export interface ItemAIU {
  id: string
  desc: string
  und: string
  cant: number
  punit: number
}

export interface AIUIn {
  cd: number
  pct_a: number
  pct_i: number
  pct_u: number

  incluir_iva: boolean
  nombre_cliente: string
  tipo_proyecto: string
  material: string
}

export interface ResultadoAIU {
  cd: number
  pct_a: number
  pct_i: number
  pct_u: number
  val_a: number
  val_i: number
  val_u: number
  val_iva: number
  logistica: number
  viaticos: number
  precio_total: number
  margen_pct: number
  nombre_cliente: string
  tipo_proyecto: string
  material: string
  incluir_iva?: boolean
  // Enriched before save
  _estado_guardado?: { aiu_items?: ItemAIU[]; nombre_cliente?: string }
  ciudad_proyecto?: string
  telefono_cliente?: string
  email_cliente?: string
  inclusiones?: string[]
  exclusiones?: string[]
}

File: backend/models/cotizacion.py
from pydantic import BaseModel
from typing import Any, Dict, Literal, List


class PiezaIn(BaseModel):
    nombre: str = ""
    largo: float
    ancho: float = 0.60
    cantidad: int = 1
    unidad_venta: Literal["ml", "m2"] = "ml"
    precio_unitario: float = 0.0


class TotalesPiezasIn(BaseModel):
    piezas: List[PiezaIn]


class MermaIn(BaseModel):
    piezas: List[PiezaIn]
    categoria: str = "Mármol"


# ── Modelos para Cotización Directa ──────────────────────────────────────────

class MaterialItem(BaseModel):
    cat: str
    ref: str
    precio_m2: float
    area_placa: float


class PiezaItem(BaseModel):
    nombre: str = ""
    ml: float
    ancho_custom: float = 0.60
    cantidad: int = 1
    categoria: str = "Mármol"
    unidad_venta: str = "ml"
    zoc_trasero: bool = False
    zoc_izq: bool = False
    zoc_der: bool = False
    altura_zocalo_cm: float = 7.0


class CotizacionDirectaIn(BaseModel):
    # Material
    categoria: str = "Mármol"
    referencia: str = ""
    precio_m2: float = 220000
    area_placa_comprada: float = 5.94
    materiales_lista: List[MaterialItem] = []
    piezas: List[PiezaItem] = []
    # Proyecto
    tipo_proyecto: str = "Meson"
    etapa_label: str = "Casa terminada (limpia)"
    nombre_cliente: str = ""
    margen_pct: float = 40.0
    dias: int = 2
    personas: int = 2
    zocalo_activo: bool = False
    zocalo_ml: float = 0.0

    # Adicionales
    adicionales_activos: bool = False
    cantidades_add: List[float] = []

    # Opciones
    incluir_iva: bool = False


class CotizacionGuardarIn(BaseModel):
    numero: str = ""
    cliente: str = ""
    resultado: Dict[str, Any]


class CotizacionAIUIn(BaseModel):
    cd: float
    pct_a: float = 2.0
    pct_i: float = 2.0
    pct_u: float = 5.0

    incluir_iva: bool = True
    nombre_cliente: str = ""
    tipo_proyecto: str = ""
    material: str = ""


class CotizacionAIUGuardarIn(BaseModel):
    numero: str = ""
    cliente: str = ""
    resultado: Dict[str, Any]
